text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,375 @@
+
+On the other hand `CARGO_TARGET_BASE_DIR` is intended to be used with several projects, possibly completely unrelated to each other. As such, it does not accept relative paths, only absolute ones. If a compelling use case is present for a relative path, it can added in the future as a backward-compatible change.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+## Setting `CARGO_TARGET_BASE_DIR`
+
+The option is similar to `CARGO_TARGET_DIR` and can be set in the same places. From less to most specific:
+
+- Through the `config.toml`:
+
+  ```toml
+  [build]
+  target-base-dir = "/absolute/path/to/target/directories"
+  ```
+
+- Through the environment variable: `CARGO_TARGET_BASE_DIR="/absolute/path/to/target/directories" cargo build`
+- Through the command line flag: `cargo build --target-base-dir /absolute/path/to/target/directories`
+
+The given path must be absolute: setting `CARGO_TARGET_BASE_DIR` to an empty or relative path is an error (when used and not instantly overriden by `CARGO_TARGET_DIR`).
+
+## Resolution order relative to `CARGO_TARGET_DIR`
+
+The resolution order favors `CARGO_TARGET_DIR` in all its forms, in the interest of both backward compatibility and allowing overriding for a singular workspace:
+
+`--target-dir` > `CARGO_TARGET_DIR` > `target-dir = ...` > `--target-base-dir` > `CARGO_TARGET_BASE_DIR` > `target-base-dir = ...`
+
+See the final section for discussion on how to make `CARGO_TARGET_BASE_DIR` the default.
+
+## Naming
+
+In the example in the previous section, using `CARGO_TARGET_BASE_DIR` with `cargo build` produces named subdirectories. The name of those is partially deterministic: it is the name of the parent directory of the workspace's `Cargo.toml` manifest and an unspecified hash of the absolute path to the workspace's root, so building `work-project/crate-1` will still use the `/cargo-caches/work-project-<hash>/debug/...` directory for a `cargo build` call.
+
+This naming scheme is chosen to be simple for people to navigate but is **not considered stable**: the hashing method (and so the hash) will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
```

Comment:
> The name of those is partially deterministic: it is the name of the parent directory of the workspace's `Cargo.toml` manifest and an unspecified hash of the absolute path to the workspace's root, so building `work-project/crate-1` will still use the `/cargo-caches/work-project-<hash>/debug/...` directory for a `cargo build` call.

and

> This naming scheme ... is **not considered stable**:

Are contradictory.

Could we under specify his further and just use "work-project-<hash>` as an example of what could be done?  The most important part is we hash the manifest path.

This will let us avoid feeling the need to hash out these implementation details within the RFC and allow us to better explore what, between this and "cargo script", what approach we should take.

Reply:
Done, good point